src/urllib3/multipart/encoder.py

```diff
@@ -0,0 +1,676 @@
+
+    def read(self, length: int = -1) -> bytes:
+        return self.fd.read(length)
+
+
+PartTuples = typing.Union[
+    typing.Tuple[
+        str,
+        typing.Union[bytes, str, typing.BinaryIO],
+    ],
+    typing.Tuple[
+        str,
+        typing.Union[bytes, str, typing.BinaryIO],
+        typing.Union[bytes, str],
+    ],
+    typing.Tuple[
+        str,
+        typing.Union[bytes, str, typing.BinaryIO],
+        typing.Union[bytes, str],
+        typing.Mapping[str, str],
+    ],
+]
+Fields = typing.Union[
+    typing.Mapping[str, typing.Union[bytes, str, PartTuples, typing.BinaryIO]],
+    typing.Sequence[
+        typing.Tuple[str, typing.Union[bytes, str, PartTuples, typing.BinaryIO]]
+    ],
+]
+
+
+class MultipartEncoder:
+    """A memory-efficient way of streaming large files in multipart/form-data format.
+
+    The basic usage is:
+
+    .. code-block:: python
+
+        import urllib3
+        from urllib3.multipart import MultipartEncoder
+
+        pm = urllib3.PoolManager()
+        encoder = MultipartEncoder({'field': 'value',
+                                    'other_field', 'other_value'})
+        r = pm.urlopen(
+            method='POST',
+            url='https://httpbin.org/post',
+            body=encoder,
+            headers=encoder.request_headers,
+        )
+
+    If you do not need to take advantage of streaming the post body, you can
+    also do:
+
+    .. code-block:: python
+
+        import urllib3
+        from urllib3.multipart import MultipartEncoder
+
+        pm = urllib3.PoolManager()
+        encoder = MultipartEncoder({'field': 'value',
+                                    'other_field', 'other_value'})
+        r = pm.urlopen(
+            method='POST',
+            url='https://httpbin.org/post',
+            body=encoder.getall(),
+            headers=encoder.request_headers,
+        )
+
+    If you want the encoder to use a specific order, you can use an
+    OrderedDict or more simply, a list of tuples:
+
+    .. code-block:: python
+
+        encoder = MultipartEncoder([('field', 'value'),
+                                    ('other_field', 'other_value')])
+
+    You can also provide tuples as part values in the same formats as are
+    supported by :meth:`~urllib3.fields.RequestField.from_tuples`
+
+    .. code-block:: python
+
+        encoder = MultipartEncoder({
+            'field': ('file_name', b'{"a": "b"}', 'application/json',
+                      {'X-My-Header': 'my-value'})
+        ])
+
+    Finally, you can also optionally specify the boundary string to use.
+    """
+
+    def __init__(
+        self,
+        fields: Fields,
+        boundary: typing.Optional[str] = None,
+        encoding: str = "utf-8",
+    ):
+        #: Boundary value either passed in by the user or created
+        self.boundary_value: str = boundary or uuid.uuid4().hex
+
+        # Computed boundary
+        self.boundary: str = f"--{self.boundary_value}"
```

Comment:
Generally speaking, this isn't there for the users, but we can make these private/read-only as necessary. It's just more to write everything we use frequently as `_{name}` because we don't trust users to be intelligent and not muck with things.

Reply:
Yeah, users will be users. :shrug: At least if these attributes are private we can say "told you so" if it ends up being a problem later.